opentech/apply/review/views.py

```diff
@@ -30,32 +34,48 @@ def get_context_data(self, **kwargs):
         for review in reviews:
             reviews_dict[review.author.pk] = review
 
+        # Get all the authors of opinions, these authors should not show up in the 'xxx_not_reviewed' lists
+        opinion_authors = set(User.objects.filter(
+            pk__in=ReviewOpinion.objects.filter(review__submission=self.object).values('author__pk')))
+
         reviews_block = defaultdict(list)
         for assigned_reviewer in assigned:
             reviewer = assigned_reviewer.reviewer
             role = assigned_reviewer.role
             review = reviews_dict.get(reviewer.pk, None)
+            key = None
             if role:
                 if review:
                     key = 'role_reviewed'
-                else:
+                elif reviewer not in opinion_authors:
                     key = 'role_not_reviewed'
             elif reviewer.is_apply_staff:
                 if review:
                     key = 'staff_reviewed'
-                else:
+                elif review not in opinion_authors:
                     key = 'staff_not_reviewed'
             else:
                 if review:
                     key = 'external_reviewed'
-                else:
+                elif review not in opinion_authors:
                     key = 'external_not_reviewed'
-
-            reviews_block[key].append({
-                'reviewer': reviewer,
-                'review': review,
-                'role': role,
-            })
+            if key:  # Do not add this reviewer to any list if they haven't reviewed but have left an opinion
+                review_info_dict = {
+                    'reviewer': reviewer,
+                    'review': review,
+                    'role': role,
+                }
+                opinions_list = []
+                if review:
+                    # Loop through all opinions and include the current author's role for this submission
+                    for opinion in review.opinions.all():
+                        opinions_list.append({
```

Comment:
i dont think you need to prefetch this in the view. you have access to all of this information off the review object within the templates.

You can even `selected_related("author")` to reduce queries.

Reply:
you are correct, excellent point